compiler/resolution/fixupExports.cpp

```diff
@@ -0,0 +1,425 @@
+  for (size_t i = 0; i < fns.size(); i++) {
+    FnSymbol* fn = fns[i];
+    SET_LINENO(fn);
+    attemptFixups(fn);
+  }
+}
+
+void fixupExportedFunction(FnSymbol* fn) {
+  attemptFixups(fn);
+}
+
+Type* getCharPtrType(void) {
+  static Type* result = NULL;
+
+  if (result != NULL) { return result; }
+
+  const char* modname = "ExternalString";
+  const char* symname = "chpl__exportTypeCharPtr";
+  ModuleSymbol* esm = NULL;
+  Symbol* alias = NULL;
+
+  // Get a handle to the ExternalString module.
+  forv_Vec(ModuleSymbol, md, allModules) {
+    if (md->modTag == MOD_INTERNAL && !strcmp(md->name, modname)) {
+      esm = md;
+      break;
+    }
+  }
+
+  INT_ASSERT(esm != NULL);
+
+  // Get a handle to the c_ptr(c_char) type from our global alias.
+  for_alist(expr, esm->block->body) {
+    DefExpr* de = toDefExpr(expr);
+    if (de) {
+      Symbol* sym = de->sym;
+      if (!strcmp(sym->name, symname)) {
+        alias = sym;
+        break;
+      }
+    }
+  }
+
+  SymExpr* rctx = new SymExpr(alias);
+
+  // Insert into tree and resolve, just in case.
+  chpl_gen_main->insertAtHead(rctx);
+  resolveExpr(rctx);
+  rctx->remove();
+
+  INT_ASSERT(alias->type != NULL);
+  result = alias->type;
+
+  return result;
+}
+
+static void attemptFixups(FnSymbol* fn) {
+  std::vector<VarSymbol*> tmps;
+  FnSymbol* wrapper = NULL;
+
+  if (not fn->hasFlag(FLAG_EXPORT)) { return; }
+
+  for (int i = 1; i <= fn->numFormals(); i++) {
+    ArgSymbol* as = fn->getFormal(i);
+
+    if (needsFixup(as)) {
+      validateFormalIntent(fn, i);
+      if (wrapper == NULL) { wrapper = createWrapper(fn); }
+      VarSymbol* tmp = fixupFormal(wrapper, i);
+      INT_ASSERT(tmp != NULL);
+      tmps.push_back(tmp);
+    } else {
+      // Push back a NULL sentry value for unconverted formals.
+      tmps.push_back(NULL);
+    }
+  }
+
+  if (needsFixup(fn->retType)) {
+    validateReturnIntent(fn);
+    if (wrapper == NULL) { wrapper = createWrapper(fn); }
+    changeRetType(wrapper);
+  }
+
+  // If a wrapper hasn't been made yet, there's nothing to do.
+  if (wrapper == NULL) { return; }
+
+  insertUnwrappedCall(wrapper, fn, tmps);
+
+  wrapperMap[wrapper] = fn;
+
+  return;
+}
+
+static bool needsFixup(ArgSymbol* as) {
+  return needsFixup(as->type);
+}
+
+static bool needsFixup(Type* t) {
+  if (t == dtString) {
+    return true;
+  }
+
+  return false;
+}
+
+static void validateFormalIntent(FnSymbol* fn, int idx) {
+  ArgSymbol* as = fn->getFormal(idx);
+  Type* t = as->type;
+
+  //
+  // TODO: If we ever add more types to these fixup routines, we really ought
+  // to put these conditions in tables.
+  //
+  if (t == dtString) {
+    IntentTag tag = as->intent;
+
+    // TODO: After resolution, have abstract intents been normalized?
+    if (tag != INTENT_CONST &&
+        tag != INTENT_CONST_REF &&
+        tag != INTENT_BLANK) {
+      SET_LINENO(fn);
+      USR_FATAL(as, "Formal %s of type \'%s\' in exported routine \'%s\' may "
```

Comment:
Is there a test for this error message?

If this error case wouldn't prevent other functions from being resolved or transformed you could turn this into a ``USR_FATAL_CONT`` so that compilation will continue and more error messages can be issued (if they exist).